### ISSUE TITLE
Allow plugins to change method signature

### ISSUE THREAD

**Issue by kedder** [python/mypy#6162]
I'm proposing to add a hook for plugins to change the function/method signature during the semantic analysis phase. I'm willing to create a PR if we agree on approach.

Here is the motivation: I'm making a [mypy-zope plugin](https://github.com/Shoobx/mypy-zope) to handle zope interfaces. Zope interfaces are weird in a sense that they are defined as classes with methods lacking a mandatory `self` argument, like this:

```python
class ISomething(zope.interface.Interface):
    def hello():
        pass
```

Mypy emits an error when finds such a declaration: `error: Method must have at least one argument`. The full example is [here](https://github.com/Shoobx/mypy-zope/blob/master/tests/samples/no_arg_method.py).

One of the way to work around this would be to have a plugin to modify a signature to insert the `self` argument before mypy checks for it, but unfortunately there is no hook available to do that at the moment. There are `get_method_signature_hook` and `get_method_hook`, but they are called on much later stage, during type checking.

I'm proposing to introduce a new hook, e.g. `get_function_analyze_hook()` that would accept both current class type and function definition, so that it can customize the function/method. I think the reasonable place to call this hook would be in `SemanticAnalyzerPass2._visit_func_def()`  before `if self.is_class_scope()`.

What do you think?

**Comment by ilevkivskyi** [comment]
Adding a hook that allows early modification of method signature looks like a too "narrow" solution.
I would propose two more ideas for discussion:
* Add three hooks similar to current `get_base_class_hook()` and co., with the only difference that they are applied before the class body is analyzed.
* Add a hook that will allow suppressing emitted error messages.

I like the latter more, but I think we need to think more about the best solution here.

**Comment by kedder** [comment]
> Add a hook that will allow suppressing emitted error messages.

This would not be ideal, because the same code that emits an error (`prepare_method_signature`) also changes the type of the very first argument (in case it is present), which is not the case for zope.interface methods. E.g. in `def add(x, y)`, the `x` would be treated as `self` of "normal" methods, even though no error would be detected.

> Add three hooks similar to current get_base_class_hook() and co., with the only difference that they are applied before the class body is analyzed.

Hm. Actually, `get_customize_class_mro_hook` is called just before body is analyzed, so I was able to do the method fixes by carefully adjusting AST nodes. Using `get_customize_class_mro_hook` for that feels a bit like an abuse, but works for now.

Thanks for the insight! Closing this issue as it looks like plugin interface extension is not unavoidable (for this feature at least).
